src/python/pants/engine/rules.py

```diff
@@ -427,6 +427,17 @@ def __new__(cls, union_base, union_member):
     return super().__new__(cls, union_base, union_member)
 
 
+@dataclass(frozen=True)
+class UnionMembership:
+  union_rules: Dict[type, Iterable[type]]
+
+  def is_union_member(self, union_type, tgt):
+    try:
+      return type(tgt.adaptor) in self.union_rules[union_type]
+    except KeyError:
```

Comment:
Would it be cleaner to explicitly check and raise? Probably this is the only source of a `KeyError` here, but using `get` fully avoids ambiugity.